### PR TITLE
[GLUTEN-11355][UT] Add new Spark 4.1 tests

### LINKED ISSUE #11355: Add new spark 4.1.x suite
### Description

Add new spark 4.1 suite 

### Gluten version

None

### PR DESCRIPTION
**Pull request by baibaichen**
This PR automates the migration of Gluten test and  SQL Test Suites from spark 4.1
Fixes #11355

## How to Generate Spark 4.1 Gluten Test Suites

With the help of AI, we follow a four-step process—details can be found [here](https://github.com/baibaichen/Blog2/blob/master/Gluten/Using%20AI%20Create%20New%20Suite/%E8%B1%86%E5%8C%85-%E5%88%9B%E5%BB%BA%E6%96%B0%20Suite%20%E7%9A%84%E6%8C%87%E4%BB%A4.md)

### 1 Baseline Analysis
- Scan existing Gluten Spark 4.0 test suites (`Gluten*Suite.scala`)
- Map each Gluten suite to its corresponding Spark 4.0 source file
- Generate mapping report with found/not-found status

### 2 Package Scope Extraction
- Extract unique package paths from successfully mapped suites
- Create a focused list of directories to monitor for changes

### 3 Delta Detection
- Compare Spark 4.0 vs Spark 4.1 test files in identified packages
- Identify net-new `*Suite.scala` files introduced in Spark 4.1

### 4 Code Generation
- For each new Spark 4.1 suite, generate corresponding Gluten wrapper
- Preserve original package structure
- Auto-create directory hierarchy as needed

---

**Result**:  Automated generation of 27 Gluten test suites for gluten Spark 4.1, ensuring test coverage parity with upstream changes. 

### Three-Way Git Merge Approach

This PR migrates SQL test suites (`inputs/*. sql` and `results/*.sql.out`) from Spark 4.1 using Git's three-way merge algorithm to preserve both upstream changes and Gluten-specific modifications.

---

**Result**: 
- Auto-merged: 165 files
- New tests added: 31 files (collations, edge cases, recursion, spatial, etc.)
- Modified tests: 134 files
- Deleted tests: 2 files (collations.sql -> split into 4 files, timestamp-ntz.sql)

## Fix and Exclusion

| Cause | Type | Category | Description | Affected files |
|-------|------|----------|-------------|----------------|
| [#52406](https://github.com/apache/spark/pull/52406) | 4.1.0 | Test Exclusion | Exclude "infer shredding with mixed scale" test. Shredding behavior with mixed decimal scales changed in Spark 4.1. | gluten-ut/spark41/.../velox/VeloxTestSettings.scala<br>**Excluded test:**<br>- "infer shredding with mixed scale" in GlutenFileBasedDataSourceSuite |
| [#50599](https://github.com/apache/spark/pull/50599) | Fix | Bug Fix | Implement Kryo serialization for CachedColumnarBatch. Adds proper Kryo serialization support for columnar cache to work with Spark's Kryo-based serialization. | backends-velox/.../ColumnarCachedBatchSerializer.scala<br>gluten-ut/spark41/.../GlutenCacheTableInKryoSuite.scala |
| [#50230](https://github.com/apache/spark/pull/50230) | 4.1.0 | Test Exclusion | Exclude GlutenMapStatusEndToEndSuite and configure parallelism. MapStatus serialization behavior changed in Spark 4.1 requiring test adjustments. | gluten-ut/spark41/.../velox/VeloxTestSettings.scala<br>gluten-ut/spark41/.../GlutenMapStatusEndToEndSuite.scala<br>**Excluded tests:**<br>- GlutenMapStatusEndToEndSuite (entire suite) |
| [#52473](https://github.com/apache/spark/pull/52473)<br>[#52870](https://github.com/apache/spark/pull/52870)<br>[#52891](https://github.com/apache/spark/pull/52891) | 4.1.0 | Test Exclusion | Exclude Spark Structured Streaming tests in Gluten. Multiple streaming-related tests need fixes due to upstream Spark changes. | gluten-ut/spark41/.../velox/VeloxTestSettings.scala<br>**Excluded tests:**<br>- GlutenStreamRealTimeModeAllowlistSuite: "rtm operator allowlist", "repartition not allowed", "stateful queries not allowed"<br>- GlutenStreamRealTimeModeE2ESuite: "foreach", "to_json and from_json round-trip", "generateExec passthrough"<br>- GlutenStreamRealTimeModeSuite: "processAllAvailable" |
| - | 4.1.0 | Test Exclusion | Exclude failing SQL tests on Spark 4.1. Multiple SQL test files need to be fixed for Spark 4.1 compatibility. | gluten-ut/spark41/.../velox/VeloxSQLQueryTestSettings.scala<br>**Excluded tests:**<br>- cast.sql<br>- describe.sql<br>- nonansi/cast.sql<br>- nonansi/st-functions.sql<br>- scripting/randomly_generated_scripts.sql<br>- st-functions.sql<br>- type-coercion-edge-cases.sql<br>- variant-field-extractions.sql |
| [#50287](https://github.com/apache/spark/pull/50287) | Fix | Bug Fix | Replace `RuntimeReplaceable` with its `replacement` to fix unit tests. This handles the RuntimeReplaceable expression changes in Spark to ensure proper expression transformation. | gluten-ut/common/.../GlutenTestsTrait.scala<br>gluten-ut/spark41/.../GlutenTimeExpressionsSuite.scala<br>gluten-ut/spark41/.../shim/GlutenTestsTrait.scala |

## How was this patch tested?

Run new tests.

### CODE REVIEW

File: .github/workflows/clickhouse_be_trigger.yml [hunk 1]
```diff
@@ -34,6 +33,11 @@ on:
       - 'gluten-ras/**'
       - 'gluten-ui/**'
       - 'gluten-kafka/**'
+    paths-ignore:
+      - 'shims/spark40/**'
+      - 'shims/spark41/**'
+      - 'gluten-ut/spark40/**'
+      - 'gluten-ut/spark41/**'
```
Comment (baibaichen) [inline .github/workflows/clickhouse_be_trigger.yml]: @zzcclp @lwz9103 
Since the ClickHouse backend does not support Spark 4.0 and later, I ignored it. I also ignored clickhouse_be_trigger.yml. This should be fine.

Comment (zzcclp) [reply .github/workflows/clickhouse_be_trigger.yml]: It's OK to me.

Comment (lwz9103) [reply .github/workflows/clickhouse_be_trigger.yml]: LGTM
